crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2053,9 +2049,23 @@ impl<'src> Parser<'src> {
             })
         } else {
             let name = self.parse_identifier();
-            let bound = self
-                .eat(TokenKind::Colon)
-                .then(|| Box::new(self.parse_conditional_expression_or_higher().expr));
+
+            let bound = if self.eat(TokenKind::Colon) {
+                if self.at_expr() {
+                    Some(Box::new(self.parse_conditional_expression_or_higher().expr))
+                } else {
+                    // test_err type_param_missing_bound
```

Comment:
Are yield, etc expressions allowed in this position?

Reply:
No, I've been holding off on adding those checks everywhere as I want to implement it once. But, it's a good idea to still add the test cases to verify the solution.